File: src/main/java/pl/training/module6/typecode/before/LegacyDeploymentRequest.java
package pl.training.module6.typecode.before;

import java.util.Locale;

public record LegacyDeploymentRequest(String releaseId, String zoneCode) {
    public LegacyDeploymentRequest {
        if (releaseId == null || releaseId.isBlank()) {
            throw new IllegalArgumentException("releaseId must not be blank");
        }
        if (zoneCode == null || zoneCode.isBlank()) {
            throw new IllegalArgumentException("zoneCode must not be blank");
        }
        zoneCode = zoneCode.toUpperCase(Locale.ROOT);
        if (!zoneCode.equals("TEST") && !zoneCode.equals("PROD") && !zoneCode.equals("DR")) {
            throw new IllegalArgumentException("unknown zone code: " + zoneCode);
        }
    }

    public boolean requiresApproval() {
        return zoneCode.equals("PROD") || zoneCode.equals("DR");
    }
}

File: csharp/src/Training.Module6/TypeCode/Before/LegacyDeploymentRequest.cs
namespace Training.Module6.TypeCode.Before;

public sealed record LegacyDeploymentRequest
{
    public LegacyDeploymentRequest(string? releaseId, string? zoneCode)
    {
        if (string.IsNullOrWhiteSpace(releaseId))
        {
            throw new ArgumentException("releaseId must not be blank");
        }
        if (string.IsNullOrWhiteSpace(zoneCode))
        {
            throw new ArgumentException("zoneCode must not be blank");
        }
        zoneCode = zoneCode.ToUpperInvariant();
        if (zoneCode != "TEST" && zoneCode != "PROD" && zoneCode != "DR")
        {
            throw new ArgumentException("unknown zone code: " + zoneCode);
        }
        ReleaseId = releaseId;
        ZoneCode = zoneCode;
    }

    public string ReleaseId { get; }

    public string ZoneCode { get; }

    public bool RequiresApproval() => ZoneCode == "PROD" || ZoneCode == "DR";
}
